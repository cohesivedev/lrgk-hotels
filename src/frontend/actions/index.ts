export function createAction<T extends string, P>(type: T, payload?: P) {
    const returnAction : { type: T, payload?: P, error?: true } = { type };

    if (typeof payload !== 'undefined') {
        returnAction.payload = payload;
    }

    return returnAction;
}
