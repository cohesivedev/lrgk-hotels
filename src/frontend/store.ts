import {createStore, applyMiddleware} from 'redux';
import thunk from 'redux-thunk';
import {composeWithDevTools} from 'redux-devtools-extension';

import reducers from './reducers';

export interface IHotel {
    name: string;
    starRating: number,
    facilities: string[];
}

export type IResultsState = {
    hotels: IHotel[];
    validFilters: string[];
}

export interface IAppState {
    results: IResultsState;
}

export const DEFAULT_STATE: IAppState = {
    results: {
        hotels:       [],
        validFilters: []
    }
};

export default createStore(
    reducers,
    DEFAULT_STATE,
    composeWithDevTools(
        applyMiddleware(thunk)
    )
);
