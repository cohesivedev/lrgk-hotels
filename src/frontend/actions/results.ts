import {ActionsUnion, IActionPromise} from '../types/store';
import {IHotel} from '../store';
import {createAction} from '.';
import {fetchWithMockAPI} from '../remote';

export type ResultsAction = ActionsUnion<typeof ResultsActions>;

export enum ResultsActionTypes {
    GET_RESULTS_HOTELS = 'RESULTS/GET_HOTELS'
}

export enum ResultsFetchEndpoints {
    Hotels = '/hotels'
}

export function getHotels(): IActionPromise {
    return dispatch => {
        fetchWithMockAPI(ResultsFetchEndpoints.Hotels)
            .then(payload => dispatch(ResultsActions.getResultsHotels(payload)));
    };
}

export const ResultsActions = {
    getResultsHotels: (payload: IHotel[]) => createAction(ResultsActionTypes.GET_RESULTS_HOTELS, payload)
    // possibly other result types other than hotel here
};
