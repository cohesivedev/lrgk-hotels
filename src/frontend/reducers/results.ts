import {IHotel, IResultsState} from '../store';
import {ResultsAction, ResultsActionTypes} from '../actions/results';

export function getValidFilters(payload: IHotel[]): string[] {
    const validFilters = new Set();

    payload!
        .forEach(hotel => {
            hotel.facilities
                .forEach(f => validFilters.add(f));
        });

    return Array.from(validFilters);
}

export default function (state: IResultsState, action: ResultsAction): IResultsState {
    let resultsState = {...state};

    switch (action.type) {
        case ResultsActionTypes.GET_RESULTS_HOTELS:
            const {payload} = action;


            resultsState = {
                ...resultsState,
                hotels:       payload!,
                validFilters: getValidFilters(payload!)
            };
            break;

        default:
    }

    return resultsState;
}
