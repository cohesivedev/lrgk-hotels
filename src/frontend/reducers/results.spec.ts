import resultsReducer, {getValidFilters} from './results';
import {IHotel} from '../store';
import {ResultsActionTypes} from '../actions/results';

describe('reducers/results', () => {
    const HOTELS: IHotel[] = [
        {name: 'foo', starRating: 1, facilities: ['a', 'b', 'c']},
        {name: 'bar', starRating: 5, facilities: ['a', 'c']},
        {name: 'foobar', starRating: 2, facilities: ['x', 'y', 'z']},
        {name: 'fizz', starRating: 3, facilities: ['a', 'x', 'c']},
        {name: 'buzz', starRating: 4, facilities: ['b', 'y', 'e']}
    ];

    it('getValidFilters() should return an array representing the set of unique facilities given a collection of hotels', () => {
        const validFilters = getValidFilters(HOTELS);

        expect(validFilters.join('')).toBe('abcxyze');
        expect(validFilters.join('')).not.toBe('abcexyz');
        expect(validFilters.length).toBe(7);
    });

    it('reducing GET_RESULTS_HOTELS actions returns a state matching the snapshot', () => {
        const newState = resultsReducer(
            {hotels: [], validFilters: []},
            {type: ResultsActionTypes.GET_RESULTS_HOTELS, payload: HOTELS}
        );

        expect(newState).toMatchSnapshot();
    });

    it('reducing an unrecognized action returns the state unchanged', () => {
        const oldState = {hotels: [], validFilters: []};

        const newState = resultsReducer(
            oldState,
            {type: 'YELL/LOUDLY' as any, payload: 'WOOOOOOOO' as any}
        );

        expect(newState).toMatchObject(oldState);
    });
});
