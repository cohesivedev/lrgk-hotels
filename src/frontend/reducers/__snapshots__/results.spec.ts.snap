// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/results reducing GET_RESULTS_HOTELS actions returns a state matching the snapshot 1`] = `
Object {
  "hotels": Array [
    Object {
      "facilities": Array [
        "a",
        "b",
        "c",
      ],
      "name": "foo",
      "starRating": 1,
    },
    Object {
      "facilities": Array [
        "a",
        "c",
      ],
      "name": "bar",
      "starRating": 5,
    },
    Object {
      "facilities": Array [
        "x",
        "y",
        "z",
      ],
      "name": "foobar",
      "starRating": 2,
    },
    Object {
      "facilities": Array [
        "a",
        "x",
        "c",
      ],
      "name": "fizz",
      "starRating": 3,
    },
    Object {
      "facilities": Array [
        "b",
        "y",
        "e",
      ],
      "name": "buzz",
      "starRating": 4,
    },
  ],
  "validFilters": Array [
    "a",
    "b",
    "c",
    "x",
    "y",
    "z",
    "e",
  ],
}
`;
